backend/app/modules/multisig_signatures/service/offchain.py

```diff
@@ -102,17 +107,38 @@ def save_pending_signature(
         signature_data: dict,
         user_ip: str,
     ):
-        verifier = self.verifiers[op_type]
-        if not verifier.verify_logic(
-            context, user_address=user_address, **signature_data
-        ):
-            raise InvalidMultisigSignatureRequest()
-
-        msg = json.dumps(signature_data)
-        msg_hash = hash_signable_message(
-            encode_for_signing(EncodingStandardFor.TEXT, msg)
+        message = signature_data.get("message")
+        encoding_standard = self._verify_signature(
+            context, user_address, message, op_type
         )
+        message_hash = get_message_hash(user_address, message)
+        encoded_message = prepare_encoded_message(message, op_type, encoding_standard)
+        safe_message_hash = hash_signable_message(encoded_message)
+        msg_to_save = prepare_msg_to_save(message, op_type)
+
         database.multisig_signature.save_signature(
-            user_address, op_type, msg, msg_hash, user_ip
+            user_address, op_type, msg_to_save, message_hash, safe_message_hash, user_ip
         )
         db.session.commit()
+
+    def _verify_signature(
+        self,
+        context: Context,
+        user_address: str,
+        signature_msg: dict | str,
+        op_type: SignatureOpType,
+    ) -> EncodingStandardFor:
+        verifier = self.verifiers[op_type]
+
+        if isinstance(signature_msg, dict):
```

Comment:
don't know if it won't be more clear to use op_type (as far as i'm correct we use dict for allocation, str for tos?)

Reply:
changed